### PR TITLE
A Spring Boot server to allow get and post requests

### PR DESCRIPTION
**Pull request by jmlaurenson**

### CODE REVIEW

File: src/main/java/com/example/javaproject/Controller/matcherController.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@RestController
+public class MatcherController {
+    Matcher matcher = new Matcher();
```
Comment (daustin-scottlogic) [inline src/main/java/com/example/javaproject/Controller/matcherController.java]: So by instantiating the matcher here yourself, you've not making use of Spring's Dependency Injection capabilities. Have a look at constructor injection and see if you can figure it out

File: src/test/java/com/example/javaproject/controller/MatcherControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.net.URI;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class MatcherControllerTest {
```
Comment (daustin-scottlogic) [inline src/test/java/com/example/javaproject/controller/MatcherControllerTest.java]: So generally for unit tests, you're looking to test the smallest "unit" of code you can in isolation. For unit testing the controller, that should be testing the various methods you have annotations on, without re-testing the matcher code itself.

To that end, it would be good at looking at Mockito to create a mock of the matcher. By mocking out the matcher logic and explicitly controlling the return values of its methods, you can focus your tests on the logic of the controller class

File: src/test/java/com/example/javaproject/controller/MatcherControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    @DisplayName("Check that posting an invalid order returns a 400, bad request error")
+    void ensureThatAnInvalidOrderReturnsBadRequest() throws Exception {
+        HttpClient client = HttpClient.newBuilder().build();
+        HttpRequest request = HttpRequest.newBuilder().header("Content-Type", "application/json")
+                .POST(HttpRequest
+                        .BodyPublishers
+                        .ofString("{\"account\": 2,\"price\": \"six\",\"quantity\": 4,\"action\": \"SELL\"}"))
+                .uri(URI.create("http://localhost:8080/addOrder")).build();
+        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+        assertEquals(400, response.statusCode(), "RESPONSE SHOULD RETURN 400");
```
Comment (daustin-scottlogic) [inline src/test/java/com/example/javaproject/controller/MatcherControllerTest.java]: I would like to see some tests on the message that the 400 gives back. If it ends up being garbled spring nonsense, we can look at introducing an exception handler, which will map an exception to a more user-friendly message

File: src/test/java/com/example/javaproject/controller/MatcherControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Test
+    @DisplayName("Check that getting a sell list returns 200, OK")
+    void ensureThatNonEmptySellListReturns() throws Exception {
+        HttpClient client = HttpClient.newBuilder().build();
+        HttpRequest request = HttpRequest.newBuilder().uri(URI.create("http://localhost:8080/sellOrders")).build();
+        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+        assertEquals(200, response.statusCode(), "RESPONSE SHOULD RETURN 200");
```
Comment (daustin-scottlogic) [inline src/test/java/com/example/javaproject/controller/MatcherControllerTest.java]: Similar here, it would be good to see assertions on the body of the response. Ideally your unit tests should help reassure you that your response is coming back in the right format

**Review by daustin-scottlogic — CHANGES_REQUESTED** [review]
Looks like a good start with the controller and the tests.

You should look at using spring's dependency injection to provide an instance of the matcher to the controller class, instead of creating it yourself. Look for Constructor Injection. It's probably going to be as easy as adding an annotation to the matcher and a constructor in the Controller.

Lets look at unit testing the controller code first, which will involve mocking out the matcher. It's a good place to learn about mocking with `Mockito`. For unit tests, you can call the methods you've annotated directly and just assert on the `ResponseEntity`, instead of using a HTTP Client just yet. 

Once things are unit tested, we can look at using some of Springs stuff to do a few larger, more involved integration/end-to-end style tests, where we can check the whole flow of the app by hitting a couple of different endpoints in a row and checking the result for each point is as we expect

File: src/test/java/com/example/javaproject/controller/MatcherControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    @Test
+    @DisplayName("Check that posting an invalid order returns a 415, unsupported media type")
+    void ensureThatAnInvalidOrderThrowsError() throws Exception {
+        mockMvc.perform(post("/addOrder")
+                        .content("{\"account\": 2,\"price\": \"six\",\"quantity\": 4,\"action\": \"SELL\"}"))
+                .andExpect(status().isUnsupportedMediaType());
+    }
+
+    @Test
+    @DisplayName("Check that posting an empty order returns a 415, unsupported media type")
```
Comment (daustin-scottlogic) [inline src/test/java/com/example/javaproject/controller/MatcherControllerTest.java]: This test is more testing the absence of the media type than an empty order. You should probably also check what happens when you post an empty order but do specify `.contentType(MediaType.APPLICATION_JSON)`

File: src/test/java/com/example/javaproject/controller/MatcherControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        this.mockMvc = webAppContextSetup(webApplicationContext).build();
+    }
+
+    @Test
+    @DisplayName("Check that getting an empty sell list returns a 204 error")
+    void ensureThatEmptySellListReturnsNoContent() throws Exception {
+        MvcResult result = mockMvc.perform(get("/sellOrders")
+                        .accept(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNoContent())
+                .andReturn();
```
Comment (daustin-scottlogic) [inline src/test/java/com/example/javaproject/controller/MatcherControllerTest.java]: Instead of returning I think you can do a fluent assertion here, using something similar to (might not be exactly it):

```
.andExpect(content().string("Order list is empty"));
```

File: src/test/java/com/example/javaproject/controller/MatcherControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    @DisplayName("Check that posting a new order returns a 201, created")
+    void ensureThatAddOrderReturnsCreated() throws Exception {
+        MvcResult result = mockMvc.perform(post("/addOrder")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content("{\"account\": 2,\"price\": 6.0,\"quantity\": 4,\"action\": \"SELL\"}")
+                        .accept(MediaType.APPLICATION_JSON))
+                .andExpect(status().isCreated())
+                .andExpect(MockMvcResultMatchers.content().contentType("application/json"))
+                .andReturn();
+        assertEquals("{\"account\":2,\"price\":6.0,\"quantity\":4,\"action\":\"SELL\",\"valid\":true}", result.getResponse().getContentAsString());
```
Comment (daustin-scottlogic) [inline src/test/java/com/example/javaproject/controller/MatcherControllerTest.java]: similar to above, you can do a fluent assertion, but this time do something like
```
.andExpect(content().json("<MESSAGE_GOES_HERE>"));
```

You also have the option to check specific parts of the json if you're not interested in everything you can check parts of the json with `jsonPath`:
```
.andExpect(jsonPath("path goes here").value("expected value goes here")));
```

**Review by daustin-scottlogic — COMMENTED** [review]
I think this is mostly fit for purpose just now. I think it's important to call out that the tests you've added are more towards integration tests than unit tests, since they're starting to test how your application fits together as a whole, rather than individual units of code. Also worth noting that Mockito is a specific library, and separate from the MockMVC stuff you're using, but we'll revisit it later.

You're probably sick of me droning on about tests, so we'll leave this be for now. Possibly once you've done the authentication chapter we'll revisit integration tests and do something which tests a more complex scenario that involves making multiple endpoint calls in a single test.
